### ISSUE TITLE
[src/05-led-roulette]No loadable ELF sections were found

### ISSUE THREAD

**Issue by Kotodian** [rust-embedded/discovery#452]
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/22030659/168800462-5e6d07dd-be9c-4463-ad02-775bf7912711.png">

**Comment by sirhcel** [comment]
Strange thing. This works out-of-the-box for me with 3c7bcec and Rust 1.62.0. I've modifed `Embed.toml` for the micro:bit v2
```diff
discovery/microbit/src/05-led-roulette $ git diff
diff --git a/microbit/src/05-led-roulette/Embed.toml b/microbit/src/05-led-roulette/Embed.toml
index 22375a6..d07a386 100644
--- a/microbit/src/05-led-roulette/Embed.toml
+++ b/microbit/src/05-led-roulette/Embed.toml
@@ -1,5 +1,5 @@
 [default.general]
-# chip = "nrf52833_xxAA" # uncomment this line for micro:bit V2
+chip = "nrf52833_xxAA" # uncomment this line for micro:bit V2
 # chip = "nrf51822_xxAA" # uncomment this line for micro:bit V1

 [default.reset]
```
and could qsuccessfully flash with `cargo-embed`:
```console
discovery/microbit/src/05-led-roulette $ cargo embed --features v2 --target thumbv7em-none-eabihf
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
      Config default
      Target /Users/christian/Entwicklung/rust/discovery/microbit/target/thumbv7em-none-eabihf/debug/led-roulette
     Erasing sectors ✔ [00:00:00] [################################################################################################]  4.00KiB/ 4.00KiB @ 31.42KiB/s (eta 0s )
 Programming pages   ✔ [00:00:00] [################################################################################################]  4.00KiB/ 4.00KiB @ 16.67KiB/s (eta 0s )
    Finished flashing in 0.303s
```

There is nothing interesting to see with the default code. My slightly modified solution [`examples/my-solution.rs](https://github.com/sirhcel/discovery/blob/issue-452-no-loadable-elf-sections/microbit/src/05-led-roulette/examples/my-solution.rs) makes the light go round after flashing it like
```console
discovery/microbit/src/05-led-roulette $ cargo embed --features v2 --target thumbv7em-none-eabihf --example my-solution
```
and pressing the reset button afterwards.

**Comment by Jzow** [comment]
I also had this problem at first. Then I plugged in the USB again and rebuilt the cargo clean according to the steps

**Comment by XaydBayeck** [comment]
I copy the `.cargo/config` file to my own project. And it's work... 
```shell
discovery-my on  master [?] is 📦 v0.1.0 via 🦀 v1.66.0-nightly
❯ cat .cargo/config
───────┬─────────────────────────────────────────────────
       │ File: .cargo/config
───────┼─────────────────────────────────────────────────
   1   │ [target.'cfg(all(target_arch = "arm", target_os = "none"))']
   2   │ rustflags = [
   3   │   "-C", "link-arg=-Tlink.x",
   4   │ ]
───────┴─────────────────────────────────────────────────
```

If you compare with or without this file, you will find the `Entry point address` is different by running `cargo readobj`.

Without this config file, the `Entry point address` is `0x0`. I guess this is the source of problem.

**Comment by l-0-l** [comment]
Yes, the reply mentioned above solves the problem. The reason is that the crate `cortex-m-rt`, the basic crate needed to work with this family of controllers, copies a default `link.x` file (you can search for `link.x.in` in the crate's source code) with Linker instructions to the build directory, and we must instruct the build system to use it. Furthermore, `link.x` file references a file called `memory.x`, which must be present and located somewhere on Linker's search path (`rustc-link-search`). This file should contain at least a minimal setting for the flash memory and the RAM on the target device, like:
```
MEMORY
{
  FLASH : ORIGIN = 0x00000000, LENGTH = 256K
  RAM : ORIGIN = 0x20000000, LENGTH = 16K
}
```
Further reading:
- [cortex-m-rt documentation](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/)
- [build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html)
- [lingade](https://doc.rust-lang.org/reference/linkage.html)

**Comment by Ljinzhou** [comment]
我遇到了同样的问题，我的解决方法是在“.cargo\config.toml”添加如下内容
```
[build]
target = "thumbv7em-none-eabihf"

[target.thumbv7em-none-eabihf]
runner = "probe-rs run --chip STM32F411CEUx"
rustflags = [
  "-C", "linker=rust-lld",
  "-C", "link-arg=-Tlink.x",
]
```

随后`cargo run`即可成功烧录。

AI解释的原因是：
.cargo/config.toml 缺少 rustflags 配置！

这是嵌入式 Rust 项目必须的配置，它做了两件事：

1. -C linker=rust-lld ：指定使用 rust-lld （LLD 链接器）而不是系统默认链接器。Windows 上默认链接器（MSVC link.exe）完全不知道如何处理 ARM Cortex-M 的目标文件。
2. -C link-arg=-Tlink.x ：告诉链接器使用 cortex-m-rt 提供的 link.x 链接脚本。这个脚本定义了向量表、 .text 、 .rodata 、 .data 、 .bss 等段的布局，并 INCLUDE memory.x 来获取 Flash/RAM 的地址和大小。
没有这两个参数时，链接器不知道把代码放到 Flash 地址（ 0x08000000 ），所有段都落在了地址 0x00000000 附近，而 probe-rs 认为这个地址不在任何有效的 Flash 区域内，所以报 "No loadable segments were found" 。

之前上一轮我错误地判断是 memory-x feature 冲突，实际上移除它对这个问题没有影响（但保留移除也是合理的，因为你已经有自己的 memory.x 和 build.rs ）。
